### PR TITLE
Malfunctioning AI Module: Core Defense Upgrade

### PR DESCRIPTION
**Pull request by gurfan**
This upgrade replaces the "Core Upgrade" and "Turret Upgrade" modules.
Costs 50 points at the moment and disables shunting, please suggest price changes or other things.

This is the Go-To module for Malf AIs that are dead-set on staying in their core. This module is a bundle of several abilities that make the AI core much harder (and hopefully more fun) for the crew to storm. My goal was the create an upgrade that would force the crew to go beyond the usual girder-next-to-turret strategy.

When purchased, does the following:

- Increases Turret Health to 200 (from 80)
- Halves turret shooting speed to 1.5 seconds (from 3 seconds)
- Causes turrets to fire pulse lasers that destroy walls and most importantly, girders.
- Turrets will fire once more at a target even after they leave the targeting area, making it more difficult to repeatedly enter/leave to core to bait turret fire.
- Grants the physical AI core immunity to fire damage (so its own turrets wont destroy it)
- Grants the AI the "Fortify" spell. (More on this below)

Fortify Spell:

protected

When toggled, a nifty blast door appears around the AI, reducing the damage taken by 3/4. However, while the shield is up, the AI will take constant damage as if they were out of power.

Picture Gallery:

lockopen
lockclose
lock

[tweak]
[content]

:cl:
 * rscadd: Added the "Core Defenses Upgrade" Malf Module, which gives the AI a suite of upgrades to their turrets and physical core.
 * rscdel: Removed the "Core Upgrade" and "Turret Upgrade" Malf Modules (Now part of Core Defense)

### CODE REVIEW

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 1]
```diff
@@ -45,29 +45,76 @@ rcd light flash thingy on matter drain
...
+
+/spell/aoe_turf/fortify_core/before_target(mob/user)
+	if(!isAI(user))
+		to_chat(user, "<span class'warning'>Only AIs can cast this spell. You shouldn't have this ability.</span>")
+		return 1
+	return
+	
+/spell/aoe_turf/fortify_core/cast(var/list/targets, var/mob/user)
+	var/mob/living/silicon/ai/A = user
+	var/obj/effect/overlay/ai_shield/shield
```
Comment (DamianX) [inline code/game/gamemodes/malfunction/Malf_Modules.dm:103]: `shield = locate(/obj/effect/overlay/ai_shield) in A.vis_contents`, get rid of the loop

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 2]
```diff
@@ -45,29 +45,76 @@ rcd light flash thingy on matter drain
...
+
+/obj/effect/overlay/ai_shield/proc/raise()
+	flick("lockdown-close", src)
+	icon_state = "lockdown"
+
+/spell/aoe_turf/fortify_core/before_target(mob/user)
+	if(!isAI(user))
+		to_chat(user, "<span class'warning'>Only AIs can cast this spell. You shouldn't have this ability.</span>")
+		return 1
+	return
```
Comment (DamianX) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: trailing return does nothing, remove it

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 3]
```diff
@@ -45,29 +45,76 @@ rcd light flash thingy on matter drain
...
+	name = "Fortify Core (Toggle)"
+	desc = "Reroutes your internal energy to a built-in blast shield within your core, greatly reducing damage taken. The shield will drain your power while active."
+	user_type = USER_TYPE_MALFAI
+	panel = MALFUNCTION
+	charge_type = Sp_RECHARGE
+	charge_max = 1 SECONDS
+	hud_state = "fortify"
+	override_base = "grey"
+	cooldown_min = 1 SECONDS
+	spell_flags = STATALLOWED
```
Comment (DamianX) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: why would the AI be able to use this while it's dead?

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 4]
```diff
@@ -45,29 +45,76 @@ rcd light flash thingy on matter drain
...
+		if(istype(content, /obj/effect/overlay/ai_shield))
+			shield = content
+			break
+	if(A.ai_flags & COREFORTIFY)
+		shield.lower()
+		sleep(7)
+		A.ai_flags &= ~COREFORTIFY
+	else
+		shield.raise()
+		sleep(7)
```
Comment (DamianX) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: sleeping here seems questionable, doesn't it allow one to spam-click the spell button before the .7 ticks have passed?

File: code/game/machinery/turrets.dm [hunk 5]
```diff
@@ -27,11 +27,14 @@
...
 	var/lastfired = 0
 	var/shot_delay = 30 //3 seconds between shots
+	var/fire_twice = 0
+
 	use_power = 1
 	idle_power_usage = 50
 	active_power_usage = 300
 //	var/list/targets
 	var/atom/movable/cur_target
+	var/atom/movable/prev_target
```
Comment (DamianX) [inline code/game/machinery/turrets.dm]: Looks like this is never cleared, which will prevent the target from soft-deling.

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 6]
```diff
@@ -45,29 +45,76 @@ rcd light flash thingy on matter drain
...
+		A.ai_flags |= COREFORTIFY
+	to_chat(user, "<span class='notice'>You [A.ai_flags & COREFORTIFY ? "forfity" : "unfortify"] your core.</span>")
```
Comment (DamianX) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]:
```suggestion
	to_chat(user, "<span class='notice'>You [A.ai_flags & COREFORTIFY ? "fortify" : "unfortify"] your core.</span>")
```

**Review by DamianX — COMMENTED** [review]
Remove AI

**Comment by BomberBro** [comment]
Way too cheap for the buffs it gives, the old one was 50 for much less and this one is 30 for something much, much stronger. Either make it unable to shunt since the entire theme around this is being a tanky defensive AI or up the cost. 
> Grants the physical AI core immunity to fire damage (so its own turrets wont destroy it)

Wouldn't this also make it immune to lasers from crew, or even welding tools?

**Comment by Irradiation** [comment]
Considering the amount of shit you get out of this purchase, a more appropriate price-tag should be 100 points. Or split the abilities into different costs because holy shit this is a LOT of nifty abilities for 30 points.

>make the AI core much harder (and hopefully more fun) for the crew to storm
I don't think anybody breaching the AI core ever had fun doing it, this is going to make it a worse chore overall.

**Comment by gurfan** [comment]
@BomberBro  @Irradiation 
The Module will now disable shunting when bought, raised price to 50.

File: code/game/machinery/turrets.dm (hunk 5 above)
Comment (gurfan) [reply code/game/machinery/turrets.dm]: I forgot to delete this variable, it doesnt do anything.

**Comment by gurfan** [comment]
@BomberBro 
> Wouldn't this also make it immune to lasers from crew, or even welding tools?

Yes. I think this immunity was the original purpose of the "Core Upgrade" Module but because all damage is converted to brute for silicons it never worked. I fixed it and kept it in mainly because the pulse turrets could easily be used to destroy the AI, but if total immunity to burn damage is too strong I could create a new type of pulse laser that won't damage AIs and have the turrets fire that.

**Comment by Irradiation** [comment]
Considering the vast majority of AIs and silicons get merced by lasers, making the AI immune is pretty bullshit. That and giving PULSE lasers to the AI. I guess NT doesn't care about making their stuff Deathsquad exclusive now.

**Comment by gurfan** [comment]
> Considering the vast majority of AIs and silicons get merced by lasers, making the AI immune is pretty bullshit. That and giving PULSE lasers to the AI. I guess NT doesn't care about making their stuff Deathsquad exclusive now.

I've made it so they're only resistant to fire damage while the shield from the fortify spell is up. The main reason why I gave the turrets pulse lasers was because of their wall destructing ability, but I can create a new, normal laser that has that destruction ability but without the extra damage and blue color.

**Comment by BomberBro** [comment]
I'd almost say 40 if it also prevents shunting, but overall a very fair option now since I guess you can always just move the AI core

**Comment by Blithering** [comment]
Would X-ray laser be better than pulse turrets? Still fucks over cheese strats without as much collateral.

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 7]
```diff
@@ -45,29 +45,74 @@ rcd light flash thingy on matter drain
...
+/datum/AI_Module/large/upgrade_defenses/on_purchase(mob/living/silicon/ai/user)
+	..()
+//	user.ai_flags |= COREFIRERESIST
 	for(var/obj/machinery/turret/turret in machines)
-		turret.health += 30
-		turret.shot_delay = 20
-	to_chat(user, "<span class='warning' Turrets upgraded.</span>")
+		turret.health += 120	//200 Totaldw
+		turret.shot_delay = 15
+		turret.lasertype = 3
```
Comment (jknpj) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: Magic number

**Comment by SonixApache** [comment]
conflicts
fix magic number

**Comment by gurfan** [comment]
magic number begone

**Comment by DamianX** [comment]
Why did this change the mapped turrets again?

**Comment by gurfan** [comment]
@DamianX 
nothing functionally changed about the map turrets. I adjusted turret code store a projectile type instead of a number 1-7 that would determine what's fired

**Pull request merged**